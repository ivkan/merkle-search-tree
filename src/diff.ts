// Tree difference calculation algorithm & associated types.

import { PageRange } from './page-range';
import { DiffListBuilder } from './diff-builder';

export interface DiffRange<K>
{
  start: K;
  end: K;
}

export function cloneDiffRange<K>(range: DiffRange<K>): DiffRange<K>
{
  return {
    start: range.start,
    end  : range.end,
  };
}

export function overlaps<K>(self: DiffRange<K>, p: DiffRange<K>): boolean
{
  return p.end >= self.start && p.start <= self.end;
}

export function start<K>(range: DiffRange<K>): K
{
  return range.start;
}

export function end<K>(range: DiffRange<K>): K
{
  return range.end;
}

export function diff<T, U, K>(local: T, peer: U): DiffRange<K>[]
{
  const localIter = local[Symbol.iterator]();
  const peerIter  = peer[Symbol.iterator]();

  const diffBuilder = new DiffListBuilder<K>();

  const localPeekable = new Peekable(localIter);
  const peerPeekable  = new Peekable(peerIter);

  const root = peerPeekable.peek();
  if (!root) return [];

  recurseDiff(root, peerPeekable, localPeekable, diffBuilder);

  return diffBuilder.intoDiffVec();
}

function recurseSubtree<p, a, T, U, K>(
  subtreeRoot: PageRange<K>,
  peer: Peekable<U>,
  local: Peekable<T>,
  diffBuilder: DiffListBuilder<K>
): boolean
{
  recurseDiff(subtreeRoot, peer, local, diffBuilder);

  while (peer.nextIf(v => subtreeRoot.isSupersetOf(v)))
  {
    diffBuilder.inconsistent(peer.current.start, peer.current.end);
  }

  return true;
}

function recurseDiff<p, a, T, U, K>(
  subtreeRoot: PageRange<K>,
  peer: Peekable<U>,
  local: Peekable<T>,
  diffBuilder: DiffListBuilder<K>
): void
{
  let lastP: PageRange<K>|null = null;

  while (true)
  {
    const p = maybeAdvanceWithin(subtreeRoot, peer);
    if (!p) return;

    const l = maybeAdvanceWithin(p, local);
    if (!l)
    {
      const start = lastP ? lastP.end : subtreeRoot.start;
      const end   = local.peek()?.start ?? p.end;
      if (end >= start)
      {
        diffBuilder.inconsistent(start, end);
      }
      return;
    }

    lastP = p;

    while (local.nextIf(v => v.isSupersetOf(p)))
    {
    }

    if (l.hash === p.hash)
    {
      diffBuilder.consistent(p.start, p.end);
      skipSubtree(p, peer);
    }
    else
    {
      diffBuilder.inconsistent(p.start, p.end);
    }

    recurseSubtree(p, peer, local, diffBuilder);
  }
}

function maybeAdvanceWithin<a, p, K, T>(
  parent: PageRange<K>,
  cursor: Peekable<T>
): PageRange<K>|null
{
  if (cursor.peek() && !parent.isSupersetOf(cursor.peek()))
  {
    return null;
  }
  return cursor.next();
}

function skipSubtree<p, T, K>(subtreeRoot: PageRange<K>, iter: Peekable<T>): void
{
  while (iter.nextIf(v => subtreeRoot.isSupersetOf(v)))
  {
  }
}

