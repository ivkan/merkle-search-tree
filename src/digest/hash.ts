import { createHash } from "crypto";

// A default Hasher implementation backed by SipHasher24

// A fast, non-cryptographic hash outputting 128-bit digests.
//
// This implementation is used as the default Hasher implementation, using
// the standard library Hash trait, which may produce non-portable hashes
// as described in the documentation of the Hash trait itself, and this
// crate's Hasher.
//
// Users may choose to initialise the SipHasher with seed keys if untrusted
// key/value user input is used in a tree in order to prevent chosen-hash
// collision attacks.
export class SipHasher {
  private hasher: any

  constructor() {
    this.hasher = createHash("shake128")
  }

  // Initialise a SipHasher with the provided seed key.
  //
  // All peers comparing tree hashes MUST be initialised with the same seed
  // key.
  static new(key: Uint8Array): SipHasher {
    const hasher = new SipHasher()
    hasher.hasher.update(key)
    return hasher
  }

  hash<T>(value: T): Uint8Array {
    const h = this.hasher.copy()
    h.update(Buffer.from(JSON.stringify(value)))
    return h.digest().slice(0, 16)
  }
}

// // Digest type
// type Digest = Buffer
//
// // Hasher interface
// interface Hasher<T> {
//   hash(value: T): Digest
// }
//
// export { SipHasher, Digest, Hasher }
