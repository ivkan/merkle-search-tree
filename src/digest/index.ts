// Hash function abstraction & digest types.
export * from './siphash';
export * from './trait';
export * from './types';
export * from './wrappers';



