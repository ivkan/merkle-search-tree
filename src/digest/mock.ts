import { Digest } from "./trait"

export class LevelKey<T>
{
  constructor(private key: T, private level: number)
  {
  }

  toString(): string
  {
    return this.key.toString();
  }
}

export class MockHasher
{
  hash(value: LevelKey<any>|string): Digest
  {
    if (value instanceof LevelKey)
    {
      const level = value.level;
      const iter  = Array(Math.floor(level / 2)).fill(0);

      let v: number[] = level % 2 === 1 ? [...iter, 0xF0] : iter;

      v = v.concat(Array(32 - v.length).fill(1));
      return new Digest(new Uint8Array(v));
    }
    else if (typeof value === 'string')
    {
      let v = new TextEncoder().encode(value);
      if (v.length > 32)
      {
        throw new Error('mock hash value is more than 32 bytes');
      }
      v = new Uint8Array([...v, ...Array(32 - v.length).fill(1)]);
      return new Digest(v);
    }
    throw new Error('Unsupported input type');
  }
}

export function level(digest: Digest): number
{
  let level = 0;
  for (const byte of digest.bytes)
  {
    if (byte === 0)
    {
      level += 2;
    }
    else if (byte === 0xF0)
    {
      level += 1;
      break;
    }
    else
    {
      break;
    }
  }
  return level;
}

