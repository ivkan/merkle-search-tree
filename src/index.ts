export * from './diff';
export * from './diff';
export * from './visitor';
export * from './node';
export * from './node-iter';
export * from './page';
export * from './tree';
