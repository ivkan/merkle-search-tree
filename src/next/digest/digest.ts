import { base64 } from '@scure/base';

/**
 * A variable bit length digest, output from a `Hasher` implementation.
 */
export class Digest<N extends number>
{
  static level = level;

  /**
   * Extract the number of leading 0's when expressed as base 16 digits, defining
   * the tree level the hash should reside at.
   */
  /*static level<N extends number>(v: Digest<N>): number
  {
    let out = 0;
    for (const byte of v.asBytes())
    {
      const zeroPrefix = zeroPrefixLen(byte);
      if (zeroPrefix === 2)
      {
        out += 2;
      }
      else if (zeroPrefix === 1)
      {
        return out + 1;
      }
      else
      {
        return out;
      }
    }
    return out;
  }*/

  private digest: Uint8Array;

  constructor(digest: Uint8Array, length?: N)
  {
    if (length && digest.length !== length)
    {
      throw new Error(`Digest must be ${length} bytes long`);
    }
    this.digest = digest;
  }

  /**
   * Wrap an opaque byte array in a `Digest` for type safety.
   */
  static new<N extends number>(digest: Uint8Array): Digest<N>
  {
    return new Digest<N>(digest);
  }

  /**
   * Return a reference to a fixed size digest byte array.
   */
  asBytes(): Uint8Array
  {
    return this.digest;
  }

  toString(): string
  {
    return base64.encode(this.digest);
  }
}

/**
 * Extract the number of leading 0's when expressed as base 16 digits, defining
 * the tree level the hash should reside at.
 */
export function level<N extends number>(v: Digest<N>, base: number): number
{
  let out = 0;
  for (const byte of v.asBytes())
  {
    const zeroCount = baseCountZero(byte, base);
    if (zeroCount === 2)
    {
      out += 2;
    }
    else if (zeroCount === 1)
    {
      return out + 1;
    }
    else
    {
      return out;
    }
  }
  return out;
}

function baseCountZero(v: number, base: number): number
{
  if (v === 0)
  {
    return 2;
  }

  if (v % base === 0)
  {
    return 1;
  }

  return 0;
}

/**
 * Returns the number of consecutive zero characters when `v` is represented as
 * a base16 string (evaluated LSB to MSB).
 */
/*function zeroPrefixLen(v: number): number
{
  // Implemented as a look-up table for fast calculation.
  switch (v)
  {
    case 0x00:
      return 2;
    case 0x10:
    case 0x20:
    case 0x30:
    case 0x40:
    case 0x50:
    case 0x60:
    case 0x70:
    case 0x80:
    case 0x90:
    case 0xA0:
    case 0xB0:
    case 0xC0:
    case 0xD0:
    case 0xE0:
    case 0xF0:
      return 1;
    default:
      return 0;
  }
}*/
