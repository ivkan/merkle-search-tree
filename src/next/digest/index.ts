export * from './siphash';
export * from './digest';
export * from './wrapper';
