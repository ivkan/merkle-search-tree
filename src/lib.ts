// Copyright 2023 Dominic Dwyer (dom@itsallbroken.com)
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy
// of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations
// under the License.

// Importing required modules
import * as tracing from './tracing';
import * as diff from './diff';
import * as digest from './digest';
import * as node from './node';
import * as nodeIter from './node_iter';
import * as page from './page';
import * as tree from './tree';
import * as visitor from './visitor';

// Exporting modules
export { diff, digest, visitor };
export * from './node';
export * from './page';
export * from './tree';

// Test-related imports and exports (commented out as they are conditionally included in Rust)
// import * as criterion from 'criterion';
// import * as tracingSubscriber from 'tracing-subscriber';
// import * as testAssert from './test_assert';
// import * as testUtil from './test_util';

// Note: TypeScript doesn't have direct equivalents for Rust's attribute macros.
// The functionality provided by these attributes (like documentation, linting, etc.)
// would typically be handled differently in a TypeScript project, often through
// separate tools like TSDoc, ESLint, etc.

