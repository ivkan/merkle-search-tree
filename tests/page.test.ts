import { Digest, PageDigest, ValueDigest } from '../src/digest';
import { Page, Node, splitOffLt } from '../src';

const MOCK_VALUE: ValueDigest    = ValueDigest.new(Digest.new(new Uint8Array(1).fill(0)));
const MOCK_PAGE_HASH: PageDigest = PageDigest.new(new Uint8Array(16).fill(0));

describe('Page Split Tests', () =>
{
  test('test_split_page_empty', () =>
  {
    let gtePage: Page<1>|null = new Page(42, []);
    expect(() => splitOffLt(gtePage, 5)).toThrowError('No nodes in this page.');
  });

  test('test_split_page_single_node_lt', () =>
  {
    let gtePage: Page<any> = new Page(42, [Node.new(2, MOCK_VALUE, null)]);
    gtePage.treeHash       = MOCK_PAGE_HASH;

    const ltPage = splitOffLt(gtePage, 5);
    console.log({
      ltPage,
      gtePage
    });
    expect(gtePage).toBeNull();

    expect(ltPage.level).toBe(42);
    expect(ltPage.treeHash).toBe(MOCK_PAGE_HASH);
    expect(ltPage.nodes).toEqual([Node.new(2, MOCK_VALUE, null)]);
  });

  test('test_split_page_single_node_gt', () =>
  {
    let gtePage: Page<any> = new Page(42, [Node.new(2, MOCK_VALUE, null)]);
    gtePage.treeHash       = MOCK_PAGE_HASH;

    const ltPage = splitOffLt(gtePage, 1);

    console.log({
      ltPage,
      gtePage
    });

    expect(ltPage).toBeNull();

    expect(gtePage.level).toBe(42);
    expect(gtePage.treeHash).toBe(MOCK_PAGE_HASH);
    expect(gtePage.nodes).toEqual([Node.new(2, MOCK_VALUE, null)]);
  });

  test('test_split_page_single_node_gt_with_high_page_split', () =>
  {
    let highPage      = new Page(40, [
      Node.new(10, MOCK_VALUE, null),
      Node.new(15, MOCK_VALUE, null),
    ]);
    highPage.treeHash = MOCK_PAGE_HASH;

    let page      = new Page(42, [Node.new(5, MOCK_VALUE, null)]);
    page.treeHash = MOCK_PAGE_HASH;
    page.insertHighPage(highPage);

    let ltPage = splitOffLt(page, 12);
    expect(page.level).toBe(40);
    expect(page.treeHash).toBeNull();
    expect(page.nodes).toEqual([Node.new(15, MOCK_VALUE, null)]);
    expect(page.highPage).toBeNull();

    expect(ltPage.level).toBe(42);
    expect(ltPage.treeHash).toBeNull();
    expect(ltPage.nodes).toEqual([Node.new(5, MOCK_VALUE, null)]);
    expect(ltPage.highPage.nodes).toEqual([Node.new(10, MOCK_VALUE, null)]);
    expect(ltPage.highPage.treeHash).toBeNull();
  });

  test('test_split_page_single_node_gt_with_child_page_split', () =>
  {
    const child2 = new Page(40, [
      Node.new(1, MOCK_VALUE, null),
      Node.new(3, MOCK_VALUE, null),
    ]);
    const child1 = new Page(41, [Node.new(4, MOCK_VALUE, child2)]);

    let page      = new Page(42, [Node.new(5, MOCK_VALUE, child1)]);
    page.treeHash = MOCK_PAGE_HASH;

    const ltPage = splitOffLt(page, 2);

    expect(page.level).toBe(42);
    expect(page.treeHash).toBeNull();
    expect(page.nodes).toEqual([
      Node.new(
        5,
        MOCK_VALUE,
        new Page(41, [Node.new(4, MOCK_VALUE, new Page(40, [Node.new(3, MOCK_VALUE, null)]))]))
    ]);

    expect(ltPage.level).toBe(40);
    expect(ltPage.treeHash).toBeNull();
  });

  test('test_split_page_eq', () =>
  {
    let gtePage: Page<any> = new Page(42, [
      Node.new(1, MOCK_VALUE, null),
      Node.new(2, MOCK_VALUE, null),
      Node.new(4, MOCK_VALUE, null),
    ]);
    gtePage.treeHash       = MOCK_PAGE_HASH;

    const ltPage = splitOffLt(gtePage, 2);
    expect(gtePage.level).toBe(42);
    expect(gtePage.treeHash).toBeNull();
    expect(gtePage.nodes).toEqual([
      Node.new(2, MOCK_VALUE, null),
      Node.new(4, MOCK_VALUE, null)
    ]);

    expect(ltPage.level).toBe(42);
    expect(ltPage.treeHash).toBeNull();
    expect(ltPage.nodes).toEqual([Node.new(1, MOCK_VALUE, null)]);
  });

  test('test_split_page_lt', () =>
  {
    let gtePage: Page<any> = new Page(42, [
      Node.new(1, MOCK_VALUE, null),
      Node.new(2, MOCK_VALUE, null),
      Node.new(4, MOCK_VALUE, null),
    ]);
    gtePage.treeHash       = MOCK_PAGE_HASH;

    const ltPage = splitOffLt(gtePage, 3);
    expect(gtePage.level).toBe(42);
    expect(gtePage.treeHash).toBeNull();
    expect(gtePage.nodes).toEqual([Node.new(4, MOCK_VALUE, null)]);

    expect(ltPage.level).toBe(42);
    expect(ltPage.treeHash).toBeNull();
    expect(ltPage.nodes).toEqual([
      Node.new(1, MOCK_VALUE, null),
      Node.new(2, MOCK_VALUE, null),
    ]);
  });

  test('test_split_page_all_gt', () =>
  {
    let gtePage: Page<any> = new Page(42, [
      Node.new(1, MOCK_VALUE, null),
      Node.new(2, MOCK_VALUE, null),
      Node.new(4, MOCK_VALUE, null),
    ]);
    gtePage.treeHash       = MOCK_PAGE_HASH;

    const ltPage = splitOffLt(gtePage, 0);
    expect(gtePage.level).toBe(42);
    expect(gtePage.treeHash).toBe(MOCK_PAGE_HASH);
    expect(gtePage.nodes).toEqual([
      Node.new(1, MOCK_VALUE, null),
      Node.new(2, MOCK_VALUE, null),
      Node.new(4, MOCK_VALUE, null),
    ]);

    expect(ltPage).toBeNull();
  });

  test('test_split_page_all_lt', () =>
  {
    let gtePage: Page<any> = new Page(42, [
      Node.new(1, MOCK_VALUE, null),
      Node.new(2, MOCK_VALUE, null),
      Node.new(4, MOCK_VALUE, null),
    ]);
    gtePage.treeHash       = MOCK_PAGE_HASH;

    const ltPage = splitOffLt(gtePage, 10);
    expect(gtePage).toBeNull();

    expect(ltPage.level).toBe(42);
    expect(ltPage.treeHash).toBe(MOCK_PAGE_HASH);
    expect(ltPage.nodes).toEqual([
      Node.new(1, MOCK_VALUE, null),
      Node.new(2, MOCK_VALUE, null),
      Node.new(4, MOCK_VALUE, null),
    ]);
  });

  // test('test_upsert_less_than_split_child', () =>
  // {
  //   let p = new Page(1, [Node.new(4, MOCK_VALUE, null)]);
  //   p.upsert(3, 0, MOCK_VALUE);
  //   p.upsert(1, 0, MOCK_VALUE);
  //   p.upsert(2, 1, MOCK_VALUE);
  //
  //   assertTree(p);
  // });

  test('test_split_page_recursive_lt_pointer', () =>
  {
    let ltPointerPage      = new Page(52, [Node.new(86, MOCK_VALUE, null)]);
    ltPointerPage.treeHash = MOCK_PAGE_HASH;

    let root      = new Page(42, [Node.new(161, MOCK_VALUE, ltPointerPage)]);
    root.treeHash = MOCK_PAGE_HASH;

    const key = 160;

    const ltPage = splitOffLt(root, key);
    expect(ltPage.level).toBe(52);
    expect(ltPage.nodes[0].getKey()).toBe(86);
  });

  test('test_split_page_recursive_high_page', () =>
  {
    let highPage      = new Page(32, [Node.new(44, MOCK_VALUE, null)]);
    highPage.treeHash = MOCK_PAGE_HASH;

    let root      = new Page(42, [Node.new(42, MOCK_VALUE, null)]);
    root.treeHash = MOCK_PAGE_HASH;
    root.insertHighPage(highPage);

    const key = 43;

    const ltPage = splitOffLt(root, key);
    expect(ltPage.level).toBe(42);
    expect(ltPage.nodes[0].getKey()).toBe(42);

    expect(root.level).toBe(32);
    expect(root.nodes[0].getKey()).toBe(44);
  });
});

