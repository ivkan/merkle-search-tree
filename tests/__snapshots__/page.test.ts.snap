// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Split Tests test_upsert_less_than_split_child 1`] = `
"digraph g {
	page_1:2 -> page_2:head;
	page_2 [shape=record, label="<head>Level 0|<1>路|1"]
	page_1:4 -> page_3:head;
	page_3 [shape=record, label="<head>Level 0|<1>路|1"]
	root [shape=diamond style=dotted];
	root -> page_1:head
	page_1 [shape=record, label="<head>Level 1|<2>路|2|<4>路|4"]
}
"
`;
